File: library/src/main/java/com/tobiasschuerg/color/models/HSLColor.java
package com.tobiasschuerg.color.models;

import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;
import android.support.v4.graphics.ColorUtils;

/**
 * Created by Tobias Schürg on 13.08.2016.
 */

public class HSLColor extends AbstractColor<HSLColor> {

    private float hue;
    private float saturation;
    private float lightness;

    public HSLColor(float h, float s, float l) {
        hue = h;
        saturation = s;
        lightness = l;
    }

    public HSLColor(ColorModel model) {
        from(model);
    }

    public HSLColor() {

    }

    @NonNull
    @Override
    public HSLColor fromColor(@ColorInt int color) {
        float[] c = new float[3];
        ColorUtils.colorToHSL(color, c);
        return new HSLColor(c[0], c[1], c[2]);
    }

    @Override
    public int toColor() {
        return ColorUtils.HSLToColor(new float[]{hue, saturation, lightness});
    }

    @NonNull
    @Override
    public HSLColor toHSL() {
        return this;
    }

    @NonNull
    @Override
    public HSVColor toHSV() {
        return new HSVColor(0, 0, 0).fromColor(toColor());
    }

    @NonNull
    @Override
    public HSLColor from(@NonNull ColorModel other) {
        return fromColor(other.toColor());
    }

    public HSLColor setLightness(float lightness) {
        this.lightness = lightness;
        return this;
    }

    public float setLightness() {
        return lightness;
    }
}

File: color/src/main/java/com/tobiasschuerg/color/models/HSLColor.kt
package com.tobiasschuerg.color.models

import androidx.annotation.ColorInt
import androidx.core.graphics.ColorUtils

/**
 * Hue, Saturation, Lightness color model.
 *
 * @param hue The hue component (0-360 degrees)
 * @param saturation The saturation component (0.0-1.0)
 * @param lightness The lightness component (0.0-1.0)
 *
 * Created by Tobias Schürg on 13.08.2016.
 */
data class HSLColor(
    val hue: Float,
    val saturation: Float,
    val lightness: Float,
) : ColorModel {
    constructor(colorModel: ColorModel) : this(colorModel.toHslArray())

    constructor(hsl: FloatArray) : this(hsl[0], hsl[1], hsl[2])

    @ColorInt
    override fun toColor(): Int {
        return ColorUtils.HSLToColor(floatArrayOf(hue, saturation, lightness))
    }

    override fun equals(other: Any?): Boolean = colorEquals(other)

    override fun hashCode(): Int = colorHashCode()

    /**
     * Creates a copy of this HSLColor with modified values.
     */
    fun setHue(newHue: Float): HSLColor = copy(hue = newHue)

    fun setSaturation(newSaturation: Float): HSLColor = copy(saturation = newSaturation)

    fun setLightness(newLightness: Float): HSLColor = copy(lightness = newLightness)
}

private fun ColorModel.toHslArray(): FloatArray {
    val c = FloatArray(3)
    ColorUtils.colorToHSL(this.toColor(), c)
    return c
}
